integration-cli/docker_cli_cp_utils.go

```diff
@@ -0,0 +1,313 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/docker/docker/pkg/archive"
+	"github.com/go-check/check"
+)
+
+type FileType uint32
+
+const (
+	Regular FileType = iota
+	Dir
+	Symlink
+)
+
+type FileData struct {
+	filetype FileType
+	path     string
+	contents string
+}
+
+func (fd FileData) creationCommand() string {
+	var command string
+
+	switch fd.filetype {
+	case Regular:
+		// Don't overwrite the file if it already exists!
+		command = fmt.Sprintf("if [ ! -f %s ]; then echo %q > %s; fi", fd.path, fd.contents, fd.path)
+	case Dir:
+		command = fmt.Sprintf("mkdir -p %s", fd.path)
+	case Symlink:
+		command = fmt.Sprintf("ln -fs %s %s", fd.contents, fd.path)
+	}
+
+	return command
+}
+
+func mkFilesCommand(fds []FileData) string {
+	commands := make([]string, len(fds))
+
+	for i, fd := range fds {
+		commands[i] = fd.creationCommand()
+	}
+
+	return strings.Join(commands, " && ")
+}
+
+var defaultFileData = []FileData{
+	{Regular, "file1", "file1"},
+	{Regular, "file2", "file2"},
+	{Regular, "file3", "file3"},
+	{Regular, "file4", "file4"},
+	{Regular, "file5", "file5"},
+	{Regular, "file6", "file6"},
+	{Regular, "file7", "file7"},
+	{Dir, "dir1", ""},
+	{Regular, "dir1/file1-1", "file1-1"},
+	{Regular, "dir1/file1-2", "file1-2"},
+	{Dir, "dir2", ""},
+	{Regular, "dir2/file2-1", "file2-1"},
+	{Regular, "dir2/file2-2", "file2-2"},
+	{Dir, "dir3", ""},
+	{Regular, "dir3/file3-1", "file3-1"},
+	{Regular, "dir3/file3-2", "file3-2"},
+	{Dir, "dir4", ""},
+	{Regular, "dir4/file3-1", "file4-1"},
+	{Regular, "dir4/file3-2", "file4-2"},
+	{Dir, "dir5", ""},
+	{Symlink, "symlink1", "target1"},
+	{Symlink, "symlink2", "target2"},
+}
+
+func defaultMkContentCommand() string {
+	return mkFilesCommand(defaultFileData)
+}
+
+func makeTestContentInDir(c *check.C, dir string) {
+	for _, fd := range defaultFileData {
+		path := filepath.Join(dir, filepath.FromSlash(fd.path))
+		switch fd.filetype {
+		case Regular:
+			makeTestFile(c, path, fd.contents+"\n")
+		case Dir:
+			makeTestDir(c, path)
+		case Symlink:
+			makeTestSymlink(c, fd.contents, path)
+		}
+	}
+}
+
+func makeTestFile(c *check.C, path, contents string) {
+	file, err := os.Create(path)
+	if err != nil {
+		c.Fatal(err)
+	}
+	defer file.Close()
+
+	if _, err = io.Copy(file, strings.NewReader(contents)); err != nil {
+		c.Fatal(err)
+	}
+}
+
+func makeTestDir(c *check.C, path string) {
+	if err := os.Mkdir(path, os.FileMode(0777)); err != nil {
+		c.Fatal(err)
+	}
+}
+
+func makeTestSymlink(c *check.C, target, linkname string) {
+	if err := os.Symlink(target, linkname); err != nil {
+		c.Fatal(err)
+	}
+}
+
+type testContainerOptions struct {
+	addContent bool
+	readOnly   bool
+	volumes    []string
+	workDir    string
+	command    string
+}
+
+func makeTestContainer(c *check.C, options testContainerOptions) (containerID string) {
```

Comment:
While there's a lot of flexibility in here, I can't help but think that it would be easier to follow and support if we just created one image/container that had all of the variants already defined.  Then we just create the same image/container for all tests and the only thing that needs to vary are the 'docker cp' commands that we run against it. And we could create it in a non-OS specific way via Dockerfiles.  Except for symlinks - that might be an issue :-)